pkg/mappings.bzl

```diff
@@ -426,3 +428,166 @@ pkg_mkdirs = rule(
     },
     provides = [PackageDirsInfo],
 )
+
+def _pkg_mklink_impl(ctx):
+    out_attributes = json.decode(ctx.attr.attributes)
+
+    # The least surprising default mode is that of a symbolic link (0777).
+    # Permissions on symlinks typically don't matter, as the operation is
+    # typically moved to where the link is pointing.
+    out_attributes.setdefault("mode", "0777")
+    return [
+        PackageSymlinkInfo(
+            destination = ctx.attr.dest,
+            source = ctx.attr.src,
+            attributes = out_attributes,
+        ),
+    ]
+
+pkg_mklink = rule(
+    doc = """Define a symlink  within packages
+
+    This rule results in the creation of a single link within a package.
+
+    Symbolic links specified by this rule may point at files/directories outside of the
+    package, or otherwise left dangling.
+
+    """,
+    implementation = _pkg_mklink_impl,
+    # @unsorted-dict-items
+    attrs = {
+        "dest": attr.string(
+            doc = """Link "target", a path within the package.
+
+            This is the actual created symbolic link.
+
+            If the directory structure provided by this attribute is not
+            otherwise created when exist within the package when it is built, it
+            will be created implicitly, much like with `pkg_files`.
+
+            This path may be prefixed or rooted by grouping or packaging rules.
+
+            """,
+            mandatory = True,
+        ),
+        "src": attr.string(
```

Comment:
How are we going to point to a target with a dynamic name? 
foo_tool ->  foo_tool_1.2-x86-glibc2.1

Reply:
`src` can obviously be a template, however there should be an option to refer to a label.

Picking up names via templates?  That's trickier.  My immediate thought was that we could pass in a label that refers to a mapping provider (e.g. `PackageFilesInfo`), along with something used to key into it, if even needed:

- Source files must be unique within a single `pkg_files` provider, so we can key on a label
- Other symbolic links are simply a one-to-one mapping, so no keying is needed
- Directories are lists of untagged entities.  Not sure about this one.  If there's only one directory, it'll Just Work(TM), but that may not be a practical ask

Another option would be to create a `links` attribute associated with mapping rules (`pkg_files`), and then the individual rules could return a provider that represents symbolic links.  This is not supported by the provider we have currently have, though.  It would also make setting attributes difficult if not impossible.

I'd need to think on this a bit more to figure out how to tie all this together, but I think that we wouldn't have to change too much about what we already have.

Reply:
I actually didn't mean using a template. I was talking about refering to a label.  If we do the label, we can pick up the complex file name easily.

Reply:
> I actually didn't mean using a template.

Ah, yes.  Regardless template via `PackageVariablesInfo` might be a good option for a lightweight case, or for pointing to dynamic locations outside of the package.

> I was talking about refering to a label. If we do the label, we can pick up the complex file name easily.

Fair enough.  However, what I was alluding to above was that our current provider interfaces may be enough to identify what the actual destination is without additional help:

1) If the provider (from rules_pkg or otherwise) points to multiple potential in-package destinations, the user will need a way to pick the precise one they want

2) If the non-packaging provider (`DefaultInfo`) points to a single source location (source/generated file), the destination may be ambiguous when we get to the grouping/packaging level.  The source location could correspond to multiple in-package destinations.

    The actual destination could be clarified by providing an additional label containing a `PackageFilesInfo` provider: sources within a single `PackageFilesInfo` must be unique.

    This doesn't solve the problem for `PackageDirsInfo`, which can contain multiple destinations with no source.

3) If a packaging provider contains a single destination (like `pkg_mklink`, or `pkg_files` and `pkg_mkdirs` with a single entry) there is no ambiguity.  The rule can calculate the destination based on what's in the provider, but then there could be additional ambiguity if multiple different providers are emitted by a single rule (the implementation currently allows this).

The easiest thing from the an implementation perspective is to enforce 3), but that might be inconvenient for users.

Regardless, I believe this can be added in after the fact.

Reply:
Discussion can continue in #313.